Week4

surendra1233 opened the pull request.

Commit:
What - Added Working Prototype of Hangman Game using OOP.
TODO - Make the game responsive so that it asks user to choose to end game or continue after completion of 1 word or chances.

Commit:
Added Working Prototype of Hangman Game using OOP.
TODO - Make the game responsive so that it asks user to choose to end game or continue after completion of 1 word or chances.

Commit:
Merge branch 'week4' of https://github.com/surendra240700/rippling_academy into week4

Commit:
Added interaction to the game.
game doesn't end after completion of 1 word. it asks to choose to quit or continue. based on the choice of the user, it serves another word or ends the game.

Commit:
updated typing for the variables in the file

Commit:
Added Hints
User can enter 0 when he want to take a hint. Randomly chooses an index which has not been guessed and changes it to the letter in the word. If multiple occurences of the letter, then user should guess the other occurences.

Commit:
written update logic in seperate function and fixed bugs
bugs fixed:
1. handled when input length is <1
2. max range of random selection of integer for hint has been changed to "(length of word - 1)"

Commit:
added abstraction in the functions to improve readability.

Commit:
removed redeclared typing

Commit:
Added spacing in the output to make User interaction better

Commit:
Addded new files to increase modularity

Hangman.py Contains the class Hangman
constants.py Contains the global variables
wordGenerator.py Contains the class WordGenerator
game.py Contains the game logic.

Commit:
restructured the code
Hangman.py - contains the basic class for hangman game
constants.py - contains only the constants
game.py - contains the class Game which handles the game. it calls the methods of hangman and handles the game.
unittestHangman.py - contains tests for the game

Commit:
Added Unit tests and modified files accordingly to pass the tests
unittesthangman.py contains unit tests that tests the methods of Game() and Hangman() classes.

Commit:
changed the file/folder names to snakecase

gekco commented on hangman/word_generator.py:
```diff
@@ -0,0 +1,17 @@
+        self.api: str = url
+        self.words: List[str] = []
+
+    def get_data(self) -> None:
```
get_data sounds like it will return data, could we change it to fetch data maybe ?

gekco commented on hangman/word_generator.py:
```diff
@@ -0,0 +1,17 @@
+        self.words: List[str] = []
+
+    def get_data(self) -> None:
+        data: str = get(self.api).text
```
also this is calling an API, have you considered what if this API fails, this generates a dependancy of your product on a third party service, if that goes down  your system stops working, but that might be unavoidable how do you handle these errors gracefully?

gekco commented on hangman/word_generator.py:
```diff
@@ -0,0 +1,17 @@
+class WordGenerator:
+    def __init__(self, url) -> None:
+        self.api: str = url
+        self.words: List[str] = []
```
also you are maintaining a list of these words in memory, for now this will work, what if we had to keep million words, how would you make changes to your design then /?

gekco commented on hangman/game.py:
```diff
@@ -0,0 +1,83 @@
+        inp = self.get_endgame_input()
+        end_game: bool = self.check_endgame(inp)
+        if(not end_game):
+            self.start_game()
```
ooops!!! recursion here is risky , can you guess why ?

gekco commented on hangman/Hangman.py:
```diff
@@ -0,0 +1,63 @@
+            print(self.hints_left, "Hints left!!!")
+            return False
+
+        while(True):
```
the logic is flawed no ? we need to show all places for that characters that we show hint for

surendra1233 replied:
Manish mentioned that we have to show only one place for that character so that user should still guess if there are any other occurrences of that character.

Commit:
Modified function name get_data to fetch_data and changed the approach from recursive to iterative
In game.py, i previously used recursive approach to restart the game. I have modified that to iterative approach in main function.

Commit:
modified code to remove redeclaration of variable

Commit:
removed unused imports

Commit:
removed unused imports

sukhdeep4395 commented on hangman/game.py:
```diff
@@ -0,0 +1,82 @@
+    def __init__(self):
+        self.word_generator = WordGenerator(SOWPODS)
+        self.word_generator.fetch_data()
+        self.hangman = None
```
hangman object can be injected via constructor. This makes unit testing easier when there are lots of dependencies for a class. 
then we don't need to initialise each of the dependency. we can just mock

also if there's a variant of hangman in future, lets say Hangman2, you can inject its object as well without changing the Game code
open close principle

sukhdeep4395 commented on hangman/game.py:
```diff
@@ -0,0 +1,82 @@
+from hangman import Hangman
+
+
+class Game:
```
we can rename Game to something more intuitive, like GameRunner.

Commit:
Renamed Game Class to GameRunner Class

sukhdeep4395 commented on hangman/game.py:
```diff
@@ -0,0 +1,82 @@
+
+    def start_game(self) -> bool:
+        word: str = self.word_generator.get_word()
+        self.hangman = Hangman(word, HINTS_LEFT, CHANCES_LEFT)
```
these two lines are initialising game, we can abstract it out for better readability
self.initialize_game()

Commit:
added typing

Commit:
moved game initialization code to game_initialize() method
